File: web/js/update.js
// web/js/update.js
//
// The update button. It exists only in the fleetdeck window: cmd/fleetdeck-
// window binds window.fleetdeckUpdate, and a browser tab has nothing to run
// an update with. The window does the work -- gets the new app, either by
// building the source tree beside it or by downloading a release and checking
// who signed it, then starts it and lets the new window put itself in place --
// and reports each step back through window.fleetdeckUpdateProgress. This file
// is what the person sees of it.
//
// The operator's rules for the button, each pinned in web/tests/update.test.js:
// it changes on the first press; a second update does not start while one
// runs (the window also holds a lock, for a second window or a terminal);
// any wait longer than two seconds shows how long it has been; and text typed
// and not sent is not lost without a word -- an update ends with the page
// reloading into the new version.
//
// One rule was added on 2026-09-12, and it is why this file changed: a build
// that cannot update itself says so. It used to say nothing, because the page
// drew a button only where the window had bound one -- so an app installed
// from a release had no button, and nobody who had one could find out that
// updating existed at all. The button is always drawn now; window.
// fleetdeckUpdateWay is what the page asks, as it loads, about whether this
// build can use it, and a build that cannot answers with the reason.

import { t } from "./i18n.js";

export const UPDATE_BINDING = "fleetdeckUpdate";
export const WAY_BINDING = "fleetdeckUpdateWay";
export const PROGRESS_FUNCTION = "fleetdeckUpdateProgress";
export const WAIT_SHOWN_AFTER_MS = 2000;
// How often a running update is repainted: the time of a wait appears no
// later than this after it passes WAIT_SHOWN_AFTER_MS.
export const UPDATE_REPAINT_MS = 250;

// phase: idle | cannot | confirm | running | done | current | busy | failed
export function initialState() {
  return { phase: "idle" };
}

// onPress returns the state after a press and whether the update is to start.
export function onPress(state, { unsent, now }) {
  switch (state.phase) {
    case "running":
    case "done":
    // A build that cannot update starts nothing, whatever reaches the button:
    // a keyboard, or a click the browser delivered before the window answered.
    case "cannot":
      return { state, start: false };
    case "confirm":
      return { state: running("press", "", now), start: true };
    default:
      if (unsent) return { state: { phase: "confirm" }, start: false };
      return { state: running("press", "", now), start: true };
  }
}

function running(step, detail, now) {
  return { phase: "running", step, detail, since: now };
}

// onProgress folds one report from the window into the state.
export function onProgress(state, { step, detail = "", reason = "" }, now) {
  switch (step) {
    case "done":
      return { phase: "done", detail };
    case "current":
      return { phase: "current", detail };
    case "busy":
      return { phase: "busy" };
    case "failed":
      return { phase: "failed", detail, reason };
    // What the window answers when this build cannot update itself at all.
    case "cannot":
      return { phase: "cannot", reason };
    // The window's answer when it can. Nothing to show, and it must not be
    // taken for a step of an update that is not running.
    case "can":
      return state;
    default:
      // A new step starts its own clock: the time shown is how long this step
      // has taken, which is the wait the person is in now.
      if (state.phase === "running" && state.step === step) return state;
      return running(step, detail, now);
  }
}

const STEP_KEYS = {
  press: "update_step_press",
  check: "update_step_check",
  build: "update_step_build",
  download: "update_step_download",
  verify: "update_step_verify",
  handover: "update_step_handover",
  "handover:alive": "update_step_alive",
  "handover:panel": "update_step_panel",
  "handover:swapped": "update_step_swapped",
  "handover:done": "update_step_swapped",
};

// reasonKey is the dictionary key for one of the window's reason codes. The
// codes are written the way the Go side spells them -- "seal:wrong-team" --
// and the dictionary keys the way every other key in it is written, so the
// two are held together here rather than in a table that can be added to on
// one side only.
export function reasonKey(prefix, reason) {
  return `${prefix}_${String(reason).replaceAll(/[:-]/g, "_")}`;
}

// reasonText is the sentence for a reason code, or "" when there is none to
// say -- an older window, or a code nobody foresaw. The caller then shows the
// particulars on their own rather than an empty line.
function reasonText(prefix, reason) {
  if (!reason) return "";
  const key = reasonKey(prefix, reason);
  const text = t(key);
  // i18n.js answers a missing key with the key itself, which on screen would
  // read as a piece of the program rather than a sentence.
  return text === key ? "" : text;
}

function shortRev(rev) {
  return String(rev ?? "").slice(0, 7);
}

function escapeHTML(s) {
  return String(s ?? "")
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
}

function fill(key, values) {
  return t(key).replace(/\{(\w+)\}/g, (_, name) => values[name] ?? "");
}

export function updateHTML(state, now) {
  const button = (label, disabled) =>
    `<button type="button" class="update-button"${disabled ? " disabled" : ""}>${escapeHTML(label)}</button>`;
  const status = (text, cls = "") => `<span class="update-status ${cls}">${escapeHTML(text)}</span>`;
  let inner;
  switch (state.phase) {
    case "confirm":
      inner = button(t("update_confirm_unsent"), false);
      break;
    case "running": {
      // Two names for the same detail, because the two sources say different
      // things with it: a tree names a commit, which is shortened to be
      // readable, and a release names a version tag, which is already short
      // and must not be cut.
      let text = fill(STEP_KEYS[state.step] ?? "update_step_press", {
        rev: shortRev(state.detail),
        version: state.detail,
      });
      const waited = now - state.since;
      if (waited >= WAIT_SHOWN_AFTER_MS) {
        text += " " + fill("update_elapsed", { n: Math.floor(waited / 1000) });
      }
      inner = button(t("update_button"), true) + status(text);
      break;
    }
    case "done":
      inner = status(fill("update_done", { rev: shortRev(state.detail) }));
      break;
    case "current":
      inner = button(t("update_button"), false) + status(fill("update_current", { rev: shortRev(state.detail) }));
      break;
    case "busy":
      inner = button(t("update_button"), false) + status(t("update_busy"), "update-problem");
      break;
    case "failed": {
      // The sentence comes from the reason code, so it is in the reader's
      // language; the detail is what the system said, in whatever language it
      // says things, and it is shown because it carries the particulars a
      // person can act on -- which team signed the app, how much room is
      // missing, what could not be reached.
      const why = reasonText("update_reason", state.reason);
      const text = why
        ? fill("update_failed_because", { why, detail: state.detail })
        : fill("update_failed", { detail: state.detail });
      inner = button(t("update_button"), false) + status(text, "update-problem");
      break;
    }
    // This build cannot update itself at all. The button stays on screen and
    // stays unpressable: its being there is how a person learns that updating
    // exists, and the sentence beside it is how they learn why not here.
    case "cannot":
      inner =
        button(t("update_button"), true) +
        status(reasonText("update_cannot", state.reason) || t("update_cannot_other"), "update-problem");
      break;
    default:
      inner = button(t("update_button"), false);
  }
  return `<span class="update-control">${inner}</span>`;
}

File: internal/supervisor/update.go
package supervisor

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"time"
)

// Progress is one thing an update tells the person who pressed the button.
type Progress struct {
	// Step is "check", "current", "done", "handover", "handover:<step>", or
	// whatever the Source reports while it puts the new app in place: "build"
	// for a tree, "download" and "verify" for a release.
	Step   string
	Detail string
}

// Source is where an update gets the app it is going to install.
//
// There are two, and the difference between them is the whole of this
// feature: Tree builds the new app from a checkout on this machine, and
// ReleaseSource downloads one that was built and signed elsewhere. Update
// drives either without knowing which it has, so that the path everything
// below it takes -- the lock, the staging directory beside the installed app,
// the handover to a new window, the swap -- is one path with one set of
// tests, and not two that drift apart.
type Source interface {
	// Check says what this source could update to: a commit, a release tag,
	// or "" when there is nothing newer than what runs.
	Check(ctx context.Context) (string, error)
	// Stage puts that version into dir and returns the bundle it put there,
	// ready to be started. It reports its own steps through say, which may be
	// nil.
	Stage(ctx context.Context, dir, version string, say func(Progress)) (string, error)
}

// BundleName is what the app bundle is called: on disk, and inside the
// release archive.
const BundleName = "fleetdeck.app"

// StagingDir is where an update builds: beside the canonical bundle, never in
// place, and on the same filesystem, so that Swap can exchange the two.
func StagingDir(canonical string) string {
	return filepath.Join(filepath.Dir(canonical), ".fleetdeck-update")
}

// NewWindowLog is the file beside the handover file the new window writes to.
const NewWindowLog = "new-window.log"

// PanelIn is the panel inside an app bundle.
func PanelIn(bundle string) string {
	return filepath.Join(bundle, "Contents", "MacOS", "fleetdeck")
}

// Update is one press of the update button, as the running window does it:
// get the new app to the side of the installed one, start the new window from
// it, and wait for it to take the panel over and put itself in place. Where
// the new app comes from is the Source's business -- a checkout brought
// forward and built, or a release downloaded and checked. The running window
// never touches the canonical bundle; the new one does, and only once it has
// shown it works (see Takeover).
type Update struct {
	Source    Source
	Canonical string // the installed app bundle
	LockPath  string
	// HandoverTimeout bounds the new window's whole takeover.
	HandoverTimeout time.Duration
	// Launch starts the new window from the staged bundle, telling it where
	// the canonical bundle and the handover file are; stop ends it.
	Launch func(staged, canonical, handover string) (stop func(), err error)
	// Pause and Resume are the running window's keeper: paused just before
	// the new window takes the panel -- or it would start the old panel again
	// the moment the new window stops it -- and resumed if the handover fails.
	Pause, Resume func()
	Progress      func(Progress)
}

func (u *Update) say(step, detail string) {
	if u.Progress != nil {
		u.Progress(Progress{Step: step, Detail: detail})
	}
}

// Run runs the update. It returns ErrBusy when another update of the same
// tree is running, and nil both when the update is done and when there was
// nothing to update (Progress says which).
func (u *Update) Run(ctx context.Context) error {
	release, err := Acquire(u.LockPath)
	if err != nil {
		return err
	}
	defer release()

	u.say("check", "")
	version, err := u.Source.Check(ctx)
	if err != nil {
		return err
	}
	if version == "" {
		u.say("current", "")
		return nil
	}

	staging := StagingDir(u.Canonical)
	// What an earlier update left there -- the bundle it swapped out -- is
	// this program's own, and an older version of it. Starting that would be
	// an update backwards.
	if err := os.RemoveAll(staging); err != nil {
		return fmt.Errorf("clear %s: %w", staging, err)
	}
	if err := os.MkdirAll(staging, 0o755); err != nil {
		return fmt.Errorf("make %s: %w", staging, err)
	}
	staged, err := u.Source.Stage(ctx, staging, version, func(p Progress) { u.say(p.Step, p.Detail) })
	if err != nil {
		return err
	}
	if _, err := os.Stat(PanelIn(staged)); err != nil {
		return fmt.Errorf("the new app has no panel in %s: %w", staged, err)
	}

	u.say("handover", "")
	h := Handover{Path: filepath.Join(staging, "handover")}
	u.Pause()
	stop, err := u.Launch(staged, u.Canonical, h.Path)
	if err != nil {
		u.Resume()
		return fmt.Errorf("start the new window: %w", err)
	}
	hctx, cancel := context.WithTimeout(ctx, u.HandoverTimeout)
	last, detail, err := h.Watch(hctx, func(s Step, d string) { u.say("handover:"+string(s), d) })
	cancel()
	if last == StepDone {
		u.say("done", version)
		return nil
	}
	stop()
	u.Resume()
	// The person is shown this, and pressing again would only fail the same
	// way: it says where the new window wrote why.
	logPath := filepath.Join(staging, NewWindowLog)
	if err != nil {
		return fmt.Errorf("the new window did not finish taking over within %s: %w; its log: %s", u.HandoverTimeout, err, logPath)
	}
	return fmt.Errorf("the new window could not take over: %s; its log: %s", detail, logPath)
}

// Takeover is the new window's half of an update: stop the panel that runs,
// start its own from the staged bundle, and -- once that panel answers with
// this build -- swap the staged bundle into the canonical path and restart the
// panel from there. Every step is reported in the handover file. Until the
// swap, the canonical bundle is the old one and any failure leaves it so; the
// staged panel is stopped before a failure is reported, so the old window can
// start its panel again.
//
// So one update starts two panels, and shows in the panel's log as two
// identical "shutting down"/"listening" pairs a moment apart -- which reads
// as two crashes, and was read that way by the operator on 2026-09-12. Both
// starts are needed, and the order is not free: the first cannot be from the
// canonical path, because the old version is still there and stays there
// until this build has shown it works, and the second is what puts the panel
// on the path the app is installed at -- the only run of the canonical bundle
// from its final path before done is reported, and the path the panel then
// reports to everything that asks. docs/engineering/window-and-panel.md,
// "The panel is started twice", has the whole of it.
type Takeover struct {
	URL       string
	Handover  Handover
	Staged    string // the bundle this window runs from
	Canonical string
	Revision  string // the build this window is; its panel must report the same
	Keeper    *Keeper
	// StartKeeper runs Keeper, from the moment the old panel is stopped.
	StartKeeper func()
	// Events are the keeper's events.
	Events <-chan Event
	// Registry is LaunchServices; nil where there is none.
	Registry Registry
	// LockPath is the update lock the old window holds while it runs the update.
	LockPath string
	// Done is called once StepDone is reported.
	Done func()
	// Logf says what went wrong without failing the takeover; nil says nothing.
	Logf func(format string, args ...any)
}

// Registry is what keeps app bundles by identifier: LaunchServices on macOS.
type Registry interface {
	Forget(bundle string) error
	Register(bundle string) error
}

// Run performs the takeover.
func (t *Takeover) Run(ctx context.Context) error {
	report := func(s Step, detail string) { _ = t.Handover.Report(s, detail) }
	fail := func(err error) error {
		// The staged panel goes before the failure is reported, so the old
		// window can start the old one on a free port.
		_ = StopHolder(context.Background(), t.URL, stopGrace)
		report(StepFailed, err.Error())
		return err
	}
	report(StepAlive, "")

	if err := StopHolder(ctx, t.URL, stopGrace); err != nil {
		report(StepFailed, err.Error())
		return err
	}
	t.StartKeeper()
	if err := t.waitOwnAnswer(ctx); err != nil {
		return fail(err)
	}
	rev, err := panelRevision(ctx, t.URL)
	if err != nil {
		return fail(err)
	}
	if rev != t.Revision {
		return fail(fmt.Errorf("the new panel reports build %q, not %q", rev, t.Revision))
	}
	report(StepPanel, rev)

	if err := Swap(t.Staged, t.Canonical); err != nil {
		return fail(err)
	}
	report(StepSwapped, t.Canonical)
	t.reregister()

	t.Keeper.Restart(PanelIn(t.Canonical))
	if err := t.waitOwnAnswer(ctx); err != nil {
		// The canonical bundle is the new one now, and it is the one that just
		// answered from the staged path: the old window, starting its panel
		// again, starts this build.
		report(StepFailed, err.Error())
		return err
	}
	report(StepDone, "")
	if t.Done != nil {
		t.Done()
	}
	t.retire(ctx)
	return nil
}

func (t *Takeover) logf(format string, args ...any) {
	if t.Logf != nil {
		t.Logf(format, args...)
	}
}

// reregister has LaunchServices forget the staged path, which holds the bundle
// swapped out now, and take the canonical one. This window checked in from the
// staged path when it started, and a later "open fleetdeck" would otherwise
// open the version just replaced (LaunchServices). A failure is said, not
// fatal: the update itself has worked, and the bundle is removed after done
// anyway.
func (t *Takeover) reregister() {
	if t.Registry == nil {
		return
	}
	if err := t.Registry.Forget(t.Staged); err != nil {
		t.logf("LaunchServices still knows the bundle swapped out at %s: %v", t.Staged, err)
	}
	if err := t.Registry.Register(t.Canonical); err != nil {
		t.logf("LaunchServices was not told of the installed bundle at %s: %v", t.Canonical, err)
	}
}

// retireWait bounds how long the new window waits for the old one's update to
// let go of the lock before the bundle swapped out is removed. The old window
// lets go as soon as it reads done, within one look at the handover file
// (handoverPoll); this is the bound on a window that never does.
const retireWait = time.Minute

// retire removes the bundle swapped out, once the old window's update has let
// go of the lock: until then that window is still running out of it. Nothing
// goes back to that bundle -- an update that fails does so before the swap and
// leaves the canonical bundle as it was -- and left in place it is the old
// version under the app's own identifier, which LaunchServices finds again the
// next time anything walks it. The handover file and the new window's log
// beside it stay.
func (t *Takeover) retire(ctx context.Context) {
	if t.LockPath == "" {
		t.logf("no update lock to wait on; the bundle swapped out stays at %s", t.Staged)
		return
	}
	deadline := time.Now().Add(retireWait)
	for {
		release, err := Acquire(t.LockPath)
		if err == nil {
			defer release()
			break
		}
		if !errors.Is(err, ErrBusy) || time.Now().After(deadline) {
			t.logf("the bundle swapped out stays at %s: the update lock was not free: %v", t.Staged, err)
			return
		}
		select {
		case <-ctx.Done():
			return
		case <-time.After(handoverPoll):
		}
	}
	if err := os.RemoveAll(t.Staged); err != nil {
		t.logf("the bundle swapped out could not be removed from %s: %v", t.Staged, err)
	}
}

// waitOwnAnswer waits for the keeper to say its own panel answers.
func (t *Takeover) waitOwnAnswer(ctx context.Context) error {
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case e := <-t.Events:
			switch {
			case e.State == Answering && e.Ours:
				return nil
			case e.State == Answering:
				return fmt.Errorf("a panel this window did not start answers at %s", t.URL)
			case e.State == Failed:
				return fmt.Errorf("%v\n%s", e.Err, e.LogTail)
			}
		}
	}
}

// panelRevision is the revision the panel at panelURL reports in its build
// fingerprint.
func panelRevision(ctx context.Context, panelURL string) (string, error) {
	u, err := url.Parse(panelURL)
	if err != nil {
		return "", err
	}
	u.Path = "/api/snapshot"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return "", err
	}
	resp, err := (&http.Client{Timeout: answerTimeout}).Do(req)
	if err != nil {
		return "", fmt.Errorf("ask the new panel for its build: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()
	var snap struct {
		Build *struct {
			Revision string `json:"revision"`
		} `json:"build"`
	}
	if err := json.NewDecoder(io.LimitReader(resp.Body, 8<<20)).Decode(&snap); err != nil || snap.Build == nil {
		return "", fmt.Errorf("the new panel's snapshot carries no build (%s)", resp.Status)
	}
	return snap.Build.Revision, nil
}
